posix: Clean up various headers

pfalcon opened the pull request:
Ok, so this started as just moving open() from unistd.h to fcntl.h, but as usual, led to chain of conflicts, which however this time were tracked down to completion.

---
Per POSIX, open() is defined in <fcntl.h>. fcntl.h in turn comes from
the underlying libc, either newlib, or minimal libc.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
posix: unistd.h: open() doesn't belong here

Per POSIX, open() is defined in <fcntl.h>. fcntl.h in turn comes from
the underlying libc, either newlib, or minimal libc.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
posix: Move sys/stat.h to minimal libc

Newlib libc already provides sys/stat.h, so trying to have sys/stat.h
on the level of POSIX subsys inevitable leads to include order and
definition conflicts. Instead (as most of other sys/* includes)
should come from the underlying libc.

While moving, made unrelated change of removing #include <kernel.h>,
to accommodate the change reviewers.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
posix: struct timespec: Move definition to sys/_timespec.h

POSIX subsys defines struct timespec in <time.h> (as POSIX public
API requires), but newlib defines in in sys/_timespec.h, which
inevitably leads to inclusion order and redifinition conflicts.
Follow newlib way and define it in single place, sys/_timespec.h,
which belongs to libc namespace. Thus, we move current definition
to minimal libc, and will use either minlibc's or newlib's
definition, instead of trying to redefine it.

This is similar to the introduction of sys/_timeval.h done earlier.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
posix: mqueue.h: Move O_CREAT and friends to fcntl.h

That's the header which is supposed to define them, there was even
FIXME on that in mqueue.h.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon committed:
posix: time.h: Add workaround for outdated newlib used by Xtensa

Unfortunately, Zephyr SDK 0.10.0 ships with outdated Newlib 2.0.0
(from 2015 or earlier) which lacks sys/_timespec.h header, requiring
ugly workaround of defining struct timespec inline (the whole idea
was to standardize on sys/_timespec.h header for different libc's).

This is similar to earlier workaround for struct timeval definition
introduced in a6aee9b4c85642053d. Zephyr SDK ticket for this issue
is https://github.com/zephyrproject-rtos/sdk-ng/issues/64, and it
will ve possible to remove both workarounds when Xtensa toolchain
will be upgraded to newlib version consistent with other
architectures.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
Ok, so this started as just moving open() from unistd.h to fcntl.h, but as usual, led to chain of conflicts, which however this time were tracked down to completion.

mbolivar commented on lib/libc/minimal/include/sys/stat.h:
```diff
@@ -1,9 +1,51 @@
+extern "C" {
+#endif
+
+#include <kernel.h>
```
Why keep this? It's not required for the definitions later in the header, and seems a bit of a layering violation for the C library to be pulling in Zephyr kernel headers.

pfalcon replied:
> Why keep this?

In general, to not introduce potential regressions. If the change says "move file", it should move file with minimal changes (I had to update header guards, as you imagine), instead of trying to squeeze other "helpful" stuff, which someone later may debug, pulling their hair out.

> and seems a bit of a layering violation for the C library to be pulling in Zephyr kernel headers.

Well, for an I/O related header like sys/stat.h there's nothing unusual to include headers related to OS I/O implementation (like kernel headers). Who knows, maybe something which includes this header relies later on those definitions to be available. Issue that or not, that's not what I fix here. Again, this started as a simple 1-line patch of moving 1 line from one file to another. This caused 4 levels of cascading failures, and turned out to required 5 patch to make CI pass. There's nothing more here to fix in *this PR*. (And a lot to fix in many more PRs.)

mbolivar replied:
> In general, to not introduce potential regressions. 

Sounds to me like any such code not caught by sanitycheck was buggy in the first place by using things it couldn't know for sure were included.

Why not remove it in a separate commit in the same PR and see what sanitycheck says?

mbolivar replied:
> not caught by sanitycheck

And out of tree users will have to do their own testing, of course. If I previously had code that pulled in fcntl.h and relied that to pull in zephyr's gpio.h, and someone changed things around so my build breaks, I think that's my fault, not theirs.

pfalcon replied:
> Sounds to me like any such code not caught by sanitycheck was buggy in the first place by using things it couldn't know for sure were included.

... And yet I still can't fix all the wrongs on Zephyr in one PR, instead trying to do that PR by PR. The sooner one PR merged, the sooner more fixes come in. Anyway, back to topic,  which color do we like them, bikesheds? Me likes green.

> Why not remove it in a separate commit in the same PR and see what sanitycheck says?

I can do that by personal request of yours, and with full understanding that it's outside the scope of this PR ;-).

mbolivar commented on include/posix/time.h:
```diff
@@ -10,14 +10,7 @@
-	signed int tv_nsec;
-};
-#endif
+#include <sys/_timespec.h>
```
Have you tested with the recently supported newlib-nano as well?

pfalcon replied:
> 

Have you tested with the recently supported newlib-nano as well?


Recently supported? Where? By whom?

mbolivar replied:
https://github.com/zephyrproject-rtos/zephyr/commit/472ea92e79e346ad390ebbfc1ff27a5776f4fa25

pfalcon replied:
> 472ea92

Ok, I'm supporting only what Zephyr CI supports. (Because I have to, or my patches won't have a chance to be merged, even though the right solution in various cases is to blacklist undermaintained platforms from CI and wait till their maintainers catch up with the rest of Zephyr).

mbolivar replied:
I think that's a fine answer.

mbolivar commented on include/posix/time.h:
```diff
@@ -10,7 +10,25 @@
 #endif
 
+#ifdef CONFIG_NEWLIB_LIBC
+/* Kludge to support outdated newlib version as used in SDK 0.10 for Xtensa */
```
If this is just for XTensa, why not ifndef under that arch too, so other arches for sure aren't negatively impacted by unexpected changes here?

pfalcon replied:
> If this is just for XTensa,

And what if this not only by Xtensa, and later we have another FooBarArch with the same issue? This kludge tests for the actual indication of the problem - very old Newlib version, and is consistent to previous such workaround already in the codebase.

mbolivar replied:
Then it's not just for XTensa. Perhaps the comment could be clarified by changing "for XTensa" to "for very old Newlib versions" or so?

pfalcon replied:
I hope I explained why I did like that: a) trying to, and usually go for, more general solution; b) yet explaining the motivation why ugliness like that appeared there at all.

You really think there's anything to change there (besides blacklisting Xtensa from POSIX CI tests)? Because I humbly think there's nothing to change for good, only for the purpose of bikeshedding.

mbolivar replied:
The comment says "for XTensa". While reviewing it, I wondered why it wasn't siloed away with ifdefs for that architecture only, especially to avoid unexpected changes on other architectures that are playing by the rules.

When I asked about that, you explained it's not just for XTensa. So I asked why not make the comment more accurate, which would have taken you about as much time to do as it did to write your reply.

You accused me of bikeshedding.

Have a nice day.

mbolivar reviewed (dismissed):
If the goal is a cleanup, I think at least including kernel.h from a libc header ought to be addressed, as it does fall under that umbrella.

pfalcon commented:
@mbolivar: I don't understand what you mean. Please review previous comments and history of title changes regarding the goal of this PR. And perhaps, post comments without hitting -1 button right away?

mbolivar commented:
>  post comments without hitting -1 button right away?

I posted 3 comments; what are you talking about?

pfalcon commented:
> I posted 3 comments; what are you talking about?

Ok, sorry, I didn't get them in email, and didn't see at once here.

mbolivar commented:
I'm done with this PR. Dismissing my stale review.

pabigot commented on lib/libc/minimal/include/sys/stat.h:
```diff
@@ -1,9 +1,51 @@
+extern "C" {
+#endif
+
+#include <kernel.h>
```
I'm with @mbolivar from the perspective of minimal libc headers should not be including kernel.h.  However I'll say it shouldn't have been put into the Zephyr's POSIX material in the first place, so that ship has sailed and though removing the include should happen it's not in scope for this PR.

nashif replied:
Should not be including kernel.h here

pfalcon replied:
> Should not be including kernel.h here

Again, I just moved the file around, and made minimal change to accompany that, if any. Let me try to remove that and see if it leads to cascade of unexpected CI breaks (which is very common so far with POSIX subsys stuff, and what I'm trying to fix - step by step).

pabigot commented:
I would like to see this merged soon since it's required as the basis of reworking #17003.

pfalcon commented:
@pabigot: Thanks for approval.

pfalcon commented:
Rebased to the latest master and removed `#include <kernel>` from sys/stat.h to accommodate thorough reviewers here.

pfalcon commented:
> Let me try to remove that and see if it leads to cascade of unexpected CI breaks (which is very common so far with POSIX subsys stuff, and what I'm trying to fix - step by step).

Failed. Surprise.

But this time it looks like some regression from timer (counter?) API rework. I'm curious why main CI didn't/doesn't catch it  - I also see our (Linaro's) CI fails on that. (Didn't have a chance to look closely, as we're at a sprint now.)

pfalcon commented:
~~~
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/timer/apic_timer.c:6:10: fatal error: drivers/system_timer.h: No such file or directory
 #include <drivers/system_timer.h>
~~~

pfalcon commented:
Build is green, @nashif, please re-review.

pabigot commented:
@pfalcon this needs a rebase so we can push it toward a merge.  Thanks.

pfalcon commented:
> this needs a rebase

Rebased, thanks. The build is green. (Note that I'm going on vacation for a week re: any possible future requests.)

pabigot commented:
@pfalcon Thanks.   My rework of #17003 is complete and waiting for this to be merged so I can submit those changes and DS3231 support.  I'm hoping that'll happen soon.  Enjoy your vacation.

andrewboie commented:
@nashif can you re-visit your review?

mgielda reviewed (approved):
LGTM

mgielda commented:
(I know Anas is on holiday and double checked that his remark was included, so I dismissed it - hope that's OK)

pfalcon committed:
libc: minimal: time.h: Don't (re)define struct timespec.

By the latest convention, libc's define struct timespec in
sys/_timespec.h. This is consistent with Newlib and ensures
about errors due to redefinitions.

Signed-off-by: Paul Sokolovsky <paul.sokolovsky@linaro.org>

pfalcon commented:
@mgielda, Thanks for review.


Also, the PR is rebased against master, added a commit to fix issue mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/17768#issuecomment-515015980 (cc: @PiotrZierhoffer).

galak commented on lib/libc/minimal/include/time.h:
```diff
@@ -34,10 +34,7 @@ struct tm {
-	time_t tv_sec;
-	long tv_nsec;
-};
+#include <sys/_timespec.h>
```
How does this work, where is <sys/_timespec.h> coming from?

galak replied:
(maybe the commit order is being shown oddly by github)

pabigot replied:
@galak https://github.com/isaacs/github/issues/386#issuecomment-494288396

galak commented on include/posix/time.h:
```diff
@@ -10,7 +10,25 @@
+/* Kludge to support outdated newlib version as used in SDK 0.10 for Xtensa */
+#include <newlib.h>
+
+#ifdef __NEWLIB__
```
Can you add a comment of which case is handling the old xtensa version

pfalcon replied:
> Can you add a comment of which case is handling the old xtensa version

Done.

galak commented on lib/libc/minimal/include/sys/stat.h:
```diff
@@ -1,9 +1,49 @@
 /*
- * Copyright (c) 2018 Linaro Limited
+ * Copyright (c) 2018 Intel Corporation
```
I'd keep the Intel Copyright and just add the Linaro one.

pfalcon replied:
> I'd keep the Intel Copyright and just add the Linaro one.

Ok. Note that the original sys/stat.h was an empty file with just a comment "Dummy sys/stat.h to fulfill the requirements of certain libraries."

galak replied:
Sure, just felt like we are moving most of the contents of include/posix/sys/stat.h into there.

pfalcon replied:
> I'd keep the Intel Copyright and just add the Linaro one.

D'oh, so what happened here, is that we had a dummy, empty file copyrighted Linaro. It gets overwritten by a moved file copyrighted by Intel. I didn't specifically change any copyright, just one file gets completely overwritten by another.

@galak, I hope you agree that under such circumstances, it doesn't make sense to stamp Linaro's copyright there.

pfalcon replied:
Ok, so hopefully this can be closed.

galak replied:
I'm fine not having the linaro copyright.

galak reviewed (changes requested):
A few minor comments, but otherwise looks good.

pfalcon commented:
> A few minor comments, but otherwise looks good.

Should be addressed now.
